sdk/helper/testcluster/exec.go

```diff
@@ -0,0 +1,288 @@
+			if err := os.MkdirAll(opts.TmpDir, 0o700); err != nil {
+				return err
+			}
+		}
+		dc.tmpDir = opts.TmpDir
+	} else {
+		tempDir, err := ioutil.TempDir("", "vault-test-cluster-")
+		if err != nil {
+			return err
+		}
+		dc.tmpDir = tempDir
+	}
+
+	// This context is used to stop the subprocess
+	execCtx, cancel := context.WithCancel(context.Background())
+	dc.stop = func() {
+		cancel()
+		close(dc.stopCh)
+	}
+	defer func() {
+		if retErr != nil {
+			cancel()
+		}
+	}()
+
+	bin := opts.BinaryPath
+	if bin == "" {
+		bin = "vault"
+	}
+
+	clusterJsonPath := filepath.Join(dc.tmpDir, "cluster.json")
+	args := []string{"server", "-dev", "-dev-cluster-json", clusterJsonPath}
+	switch {
+	case opts.NumCores == 3:
+		args = append(args, "-dev-three-node")
+	case opts.NumCores == 1:
+		args = append(args, "-dev-tls")
+	default:
+		return fmt.Errorf("NumCores=1 and NumCores=3 are the only supported options right now")
+	}
+	cmd := exec.CommandContext(execCtx, bin, args...)
+	if opts.Logger != nil {
+		stdout, err := cmd.StdoutPipe()
+		if err != nil {
+			return err
+		}
+		go func() {
+			// TODO set bigger buffer
+			scanner := bufio.NewScanner(stdout)
+			for scanner.Scan() {
+				opts.Logger.Trace(scanner.Text())
+			}
+		}()
+		stderr, err := cmd.StderrPipe()
+		if err != nil {
+			return err
+		}
+		go func() {
+			scanner := bufio.NewScanner(stderr)
+			for scanner.Scan() {
+				opts.Logger.Trace(scanner.Text())
+			}
+		}()
+	}
+
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+
+	ctx, _ = context.WithTimeout(ctx, 5*time.Second)
+	for ctx.Err() == nil {
+		if b, err := os.ReadFile(clusterJsonPath); err == nil && len(b) > 0 {
+			var clusterJson ClusterJson
+			if err := jsonutil.DecodeJSON(b, &clusterJson); err != nil {
+				continue
+			}
+			dc.CACertPEMFile = clusterJson.CACertPath
+			dc.rootToken = clusterJson.RootToken
+			for i, node := range clusterJson.Nodes {
+				config := api.DefaultConfig()
+				config.Address = node.APIAddress
+				err := config.ConfigureTLS(&api.TLSConfig{
+					CACert: clusterJson.CACertPath,
+				})
+				if err != nil {
+					return err
+				}
+				client, err := api.NewClient(config)
+				if err != nil {
+					return err
+				}
+				client.SetToken(dc.rootToken)
+				_, err = client.Sys().ListMounts()
+				if err != nil {
+					return err
+				}
+
+				dc.ClusterNodes = append(dc.ClusterNodes, &execDevClusterNode{
+					name:   fmt.Sprintf("core-%d", i),
+					Client: client,
+				})
+			}
+			return nil
+		}
+		time.Sleep(500 * time.Millisecond)
+	}
+	return ctx.Err()
+}
+
+type execDevClusterNode struct {
+	// HostPort          string
+	name   string
+	Client *api.Client
+}
+
+var _ VaultClusterNode = &execDevClusterNode{}
+
+func (e *execDevClusterNode) Name() string {
+	return e.name
+}
+
+func (e *execDevClusterNode) APIClient() *api.Client {
+	return e.Client
+}
+
+func (e *execDevClusterNode) TLSConfig() *tls.Config {
+	return e.Client.CloneConfig().TLSConfig()
```

Comment:
❓  I noticed that TLSConfig() returns a clone of the tls.Config pointed struct. Is the reason that the api.Client's config is also cloned (_i.e._ `e.Client.CloneConfig()`), that it can take advantage of the ReadLock for the overall client config?

Reply:
No, it's because CloneConfig is the only way to get access to the config field.